### PR TITLE
Added calendar view, created excusals, and created optional builds.

### PR DESCRIPTION
**Pull request by elijah-santos25**
### Database
- Added table `optional_builds` (stores the date of the optional build as a primary key)
- Added table `excused_sessions` (stores the date of excusal and student id as primary keys)
- Added associated migration files, bumping the schema version to 9
- Added associated Ruby models

### Views
- Added calendar
- Added views to add and delete excusals and optional builds
- Added a table to `student.erb` that displays a student's excusals
- All of this is only visible to users who can edit hours

### Hours Server
- Added support for prefilling dates in various routes (using URL query parameters)
- Fixed a bug that resulted in all hours manually added being stored in the wrong time zone

### Miscellaneous
- Removed slash from `config.json` for dev configuration -- this slash caused various errors when trying to start the server
- Added gem `tzinfo-data` -- allows date formatting on Windows machines during development

### CODE REVIEW

File: db/migrations/008_create_optional_builds.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Sequel.migration do
+    change do
+        create_table(:optional_builds) do
+            DateTime :date, null: false, unique: true
```
Comment (patfair) [inline db/migrations/008_create_optional_builds.rb:4]: Industry best practice is to have an integer primary key for any table, so I'd suggest adding one here. The date can still be unique.

File: db/migrations/009_create_excused_sessions.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+Sequel.migration do
+    change do
+        create_table(:excused_sessions) do
+            DateTime :date, null: false
```
Comment (patfair) [inline db/migrations/009_create_excused_sessions.rb:4]: Same here; better to have an integer primary key; then you can get rid of the hack below.

File: db/migrations/008_create_optional_builds.rb (hunk 1 above)
Comment (patfair) [inline db/migrations/008_create_optional_builds.rb:4]: Could this be just `Date` instead of `DateTime`, for simplicity?

File: hours_server.rb [hunk 3]
```diff
@@ -316,6 +391,56 @@ def sms_response(messages)
...
+              LEFT JOIN cheesy_frc_hours.lab_sessions ON DATE(cheesy_frc_hours.lab_sessions.time_in) = build_days.build_date 
+                  AND cheesy_frc_hours.lab_sessions.student_id=cheesy_frc_hours.students.id
+                  AND !cheesy_frc_hours.lab_sessions.excluded_from_total
+              LEFT JOIN cheesy_frc_hours.excused_sessions ON cheesy_frc_hours.excused_sessions.student_id=cheesy_frc_hours.students.id
+                  AND cheesy_frc_hours.excused_sessions.date=build_days.build_date
+              ORDER BY build_date ASC
+        ), student_build_info AS (
+            SELECT 
+              COUNT(IF(required AND (!excused OR (excused AND attended)), 1, NULL)) AS required_count,
+              -- COUNT(IF(excused AND !attended, 1, NULL)) AS excused_count,
```
Comment (patfair) [inline hours_server.rb:423]: Do these comments serve any purpose or can they be removed?

File: public/css/css.css [hunk 4]
```diff
@@ -1,132 +1,252 @@
...
+}
+
+summary {
+    font-weight: bold;
+    margin: -0.5em -0.5em 0;
+    padding: 0.5em;
+}
+
+table.column-based {
+	display: table;
```
Comment (patfair) [inline public/css/css.css:49]: Inconsistent indentation; please make everything in this file the same.

File: views/calendar.erb [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+                              <a href="/schedule_optional?date=<%=build_date[:build_date]%>" class="visible-if-parent-hover"><small style="font-size: 0.65em; padding: 0px 8px;">+optional</small></a>
+                           <% end %>
+                        </th>
+                  <% end %>
+               </tr>
+            </table>
+            <table style="table-layout: fixed;" class="table table-striped table-bordered table-condensed column-based">
+               <%studentsCnt = Student.count%>
+               <%rowNum = 1%>
+               <%DB.fetch """
```
Comment (patfair) [inline views/calendar.erb:50]: If you're going to have massive inline SQL queries like this, it should be either in `hours_server.rb` or here, but not both.

**Review by patfair — COMMENTED** [review]
Looks okay, overall. Some comments about best practices and style.

**Review by patfair — APPROVED** [review]

**Comment by patfair** [comment]
@elijah-santos25 after deploying, I get the attached error. Our MySQL version is 5.6 and I think your query may be using some syntax that is only supported in later versions. Can you please see if there is a way to rewrite it to work for 5.6?

Screenshot 2023-02-11 at 12 09 48 PM

**Comment by elijah-santos25** [comment]
I'll see if I can find the issue.

**Comment by elijah-santos25** [comment]
Found the issue: MySQL 5.6 doesn't support common table expressions. I'm refactoring it now.

**Comment by elijah-santos25** [comment]
Fixed the issue (I think). Should I directly modify this repo or make another PR?

**Comment by patfair** [comment]
Please create another PR.

**Comment by elijah-santos25** [comment]
Should be fixed now, see #13

**Comment by patfair** [comment]
It appears to work now, thanks!

**Pull request merged**